### PR TITLE
feature(main):  add registry sdk

### PR DESCRIPTION
**Pull request by cuisongliu**
Signed-off-by: cuisongliu <cuisongliu@qq.com> 

1. add registry sdk 
2. add image ctl to sealctl
3. image-cri-shim using registry sdk
4. delete basic-auth in pul push build merge
5. add version in print

### CODE REVIEW

**Comment by cuisongliu** [comment]
#1314 
#1740

File: pkg/image/cmd/root.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+package cmd
+
+import (
+	"github.com/spf13/cobra"
+)
+
+var contact string
+
+func Registry(rootCmd *cobra.Command, tips string) {
```
Comment (fengxsong) [inline pkg/image/cmd/root.go:25]: Register

File: cmd/sealos/cmd/add.go [hunk 2]
```diff
@@ -57,7 +57,7 @@ func newAddCmd() *cobra.Command {
 			return nil
 		},
 		PostRun: func(cmd *cobra.Command, args []string) {
```
Comment (fengxsong) [inline cmd/sealos/cmd/add.go:59]: use `PersistentPostRun` instead.

File: cmd/image-cri-shim/cmd/root.go [hunk 3]
```diff
@@ -118,16 +69,12 @@ func Execute() {
 }
 
 func init() {
-	rootCmd.Flags().StringVarP(&server.ConfigFile, "file", "f", "", "config file top image shim")
+	rootCmd.Flags().StringVarP(&cfgFile, "file", "f", "", "config file top image shim")
```
Comment (fanux) [inline cmd/image-cri-shim/cmd/root.go:72]: config file top image shim -> image shim root config

File: cmd/image-cri-shim/cmd/root.go [hunk 4]
```diff
@@ -151,6 +98,6 @@ func run(socket string, criSocket string) {
 		close(stopCh)
 	case <-stopCh:
 	}
-	_ = os.Remove(socket)
-	logger.Info("Shutting down the image _shim")
+	_ = os.Remove(cfg.ShimSocket)
+	logger.Info("shutting down the image _shim")
```
Comment (fanux) [inline cmd/image-cri-shim/cmd/root.go:102]: image _shim -> image_shim

File: pkg/constants/contants.go [hunk 5]
```diff
@@ -48,3 +48,21 @@ func GetHomeDir() string {
...
+ /\ \:\ \ \__\ /:/\:\ \:\__\ /:/\:\ \:\__\ /:/__/    /:/__/ \:\__\ /\ \:\ \ \__\
+ \:\ \:\ \/__/ \:\~\:\ \/__/ \/__\:\/:/  / \:\  \    \:\  \ /:/  / \:\ \:\ \/__/
+  \:\ \:\__\    \:\ \:\__\        \::/  /   \:\  \    \:\  /:/  /   \:\ \:\__\
+   \:\/:/  /     \:\ \/__/        /:/  /     \:\  \    \:\/:/  /     \:\/:/  /
+    \::/  /       \:\__\         /:/  /       \:\__\    \::/  /       \::/  /
+     \/__/         \/__/         \/__/         \/__/     \/__/         \/__/
+
+                  Website :https://www.sealos.io/
+                  Address :github.com/labring/sealos
+		BuildVersion: %s
```
Comment (fanux) [inline pkg/constants/contants.go:67]: align

File: staging/src/github.com/labring/image-cri-shim/pkg/server/cri_server.go [hunk 6]
```diff
@@ -132,26 +124,34 @@ func (s *server) replaceImage(image, action string) (newImage string) {
...
+			logger.Info("image %s NotIn ShimImages", image)
+			// there's no list of image-shims, or image is not found on the list
+			reg, err := registry.NewRegistryForDomain(authDomain, auth.Username, auth.Password)
+			if err == nil {
+				tags, err := reg.Tags(preImageName)
+				if err != nil {
+					logger.Warn("list tag for %s error: %+v", preImageName, err)
+				} else {
+					if str.InList(preImageTag, tags) {
+						newImage = joinImageDomainAndName(authDomain, strings.Join([]string{preImageName, preImageTag}, ":"))
```
Comment (fanux) [inline staging/src/github.com/labring/image-cri-shim/pkg/server/cri_server.go:142]: indented too much

File: staging/src/github.com/labring/image-cri-shim/pkg/server/cri_server.go [hunk 7]
```diff
@@ -132,26 +124,34 @@ func (s *server) replaceImage(image, action string) (newImage string) {
...
+						ShimImages.Insert(preImageAllName)
+						logger.Info("begin image: %s, after image: %s, action: %s", image, newImage, action)
```
Comment (fanux) [inline staging/src/github.com/labring/image-cri-shim/pkg/server/cri_server.go:144]: begin after?

File: staging/src/github.com/labring/image-cri-shim/pkg/shim/config.go [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+	Domain      string                    `json:"-"`
+	CRIConfigs  map[string]api.AuthConfig `json:"-"`
+}
+
+func (c *Config) PreProcess() error {
+	if c.ShimSocket == "" {
+		c.ShimSocket = server.SealosShimSock
+	}
+	defer func() {
+		if c.Image != "" {
```
Comment (fanux) [inline staging/src/github.com/labring/image-cri-shim/pkg/shim/config.go:58]: if c.Image == "" return

File: staging/src/github.com/labring/image-cri-shim/pkg/shim/config.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+				images, err := img.LoadImages(c.Image)
+				if err != nil {
+					logger.Warn("load images from image dir: %v", err)
+				}
+				server.ShimImages.Insert(images...)
+				logger.Info("sync image list for image dir,sync once second is %d,data is %+v", c.Sync, images)
+			}
+		}
+	}()
+	logger.Info("shim-socket: %s", c.ShimSocket)
```
Comment (fanux) [inline staging/src/github.com/labring/image-cri-shim/pkg/shim/config.go:79]: print log in one line

Comment (cuisongliu) [reply staging/src/github.com/labring/image-cri-shim/pkg/shim/config.go:80]: this this print shim socket. after this print image socket

File: pkg/constants/contants.go (hunk 5 above)
Comment (cuisongliu) [reply pkg/constants/contants.go:67]: It's okay to indent here.

```
     ___           ___           ___           ___       ___           ___
     /\  \         /\  \         /\  \         /\__\     /\  \         /\  \
    /::\  \       /::\  \       /::\  \       /:/  /    /::\  \       /::\  \
   /:/\ \  \     /:/\:\  \     /:/\:\  \     /:/  /    /:/\:\  \     /:/\ \  \
  _\:\~\ \  \   /::\~\:\  \   /::\~\:\  \   /:/  /    /:/  \:\  \   _\:\~\ \  \
 /\ \:\ \ \__\ /:/\:\ \:\__\ /:/\:\ \:\__\ /:/__/    /:/__/ \:\__\ /\ \:\ \ \__\
 \:\ \:\ \/__/ \:\~\:\ \/__/ \/__\:\/:/  / \:\  \    \:\  \ /:/  / \:\ \:\ \/__/
  \:\ \:\__\    \:\ \:\__\        \::/  /   \:\  \    \:\  /:/  /   \:\ \:\__\
   \:\/:/  /     \:\ \/__/        /:/  /     \:\  \    \:\/:/  /     \:\/:/  /
    \::/  /       \:\__\         /:/  /       \:\__\    \::/  /       \::/  /
     \/__/         \/__/         \/__/         \/__/     \/__/         \/__/

                  Website :https://www.sealos.io/
                  Address :github.com/labring/sealos
		BuildVersion: untagged-ba5cac35
```

**Review by zzjin — APPROVED** [review]

**Pull request merged**